### PR TITLE
Rename `db.name` to `db.namespace` and generalize it

### PR DESCRIPTION
**Pull request by lmolkova**
## Changes

Deprecates `db.name` and `db.redis.database_index` in favor of `db.collection.namespace`
Removes `db.mssql.instance_name` (includes it into `db.collection.namespace`).

Open question:
where does the schema belong?
  - within `name`:  `namespace = mydb`, `name = dbo.mytable` , span name is `SELECT mydb.dbo.mytable`
    - If schema is not set in the query, then we won't have it (`namespace = mydb`  and `name = mytable` and span name is `SELECT mydb.mytable`
    - But it's a best practice to qualify table name with schema
  - within `namespace`: `namespace = mydb.dbo`, `name = dbo.mytable`,  span name is `SELECT mydb.dbo.dbo.mytable`
    - if schema is provided as a part of table name (best practice), we have duplication

The schema (at least sometimes) requires actual database call: [postgres jdbc driver]( https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java#L1671), [SQL Server JDBC driver](https://github.com/microsoft/mssql-jdbc/blob/main/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java#L7271). 
It's not a connection property per se, but a database server configuration, i.e. can be modified elsewhere and can't be cached on the client. 

So in this PR schema stays inside table name (if users provided it in the query) and is not captured inside the database name.

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] Change log entry added, according to the guidelines in [When to add a changelog entry](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#when-to-add-a-changelog-entry).
  * If your PR does not need a change log, start the PR title with `[chore]`
* [x] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

### CODE REVIEW

File: docs/attributes-registry/db.md [hunk 1]
```diff
@@ -21,16 +19,17 @@
...
 | `db.query.parameter.<key>` | string | The query parameters used in `db.query.text`, with `<key>` being the parameter name, and the attribute value being the parameter value. [3] | `someval`; `55` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
 | `db.query.text` | string | The database query being executed. | `SELECT * FROM wuser_table where username = ?`; `SET mykey "WuValue"` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
 | `db.system` | string | An identifier for the database management system (DBMS) product being used. See below for a list of well-known identifiers. | `other_sql` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
 
 **[1]:** If the collection name is parsed from the query, it SHOULD match the value provided in the query and may be qualified with the schema and database name.
 
-**[2]:** In some SQL databases, the database name to be used is called "schema name". In case there are multiple layers that could be considered for database name (e.g. Oracle instance name and schema name), the database name to be used is the more specific layer (e.g. Oracle schema name).
```
Comment (lmolkova) [inline docs/attributes-registry/db.md:33]: this was not correct. Essentially schema name is a sub-namespace inside database which has 1:1 relationship with a user who owns corresponding objects.
This SO thread provides a good idea terminology and mapping between databases - https://stackoverflow.com/questions/7942520/relationship-between-catalog-schema-user-and-database-instance

It's considered a best practice to qualify table name with schema (in terms of [correctness](https://sqlblog.org/2019/09/12/bad-habits-to-kick-avoiding-the-schema-prefix) and [performance](https://www.red-gate.com/simple-talk/blogs/why-you-should-always-use-schema-name-in-queries/)).

So having schema in the `db.collection.name`  is a better choice than adding it to `db.collection.namespace`.

If we include it into `db.collection.namespace`, we're likely to duplicate it for users that follow best practices.

File: docs/database/hbase.md [hunk 2]
```diff
@@ -17,9 +17,12 @@ described on this page.
 <!-- semconv db.hbase(full,tag=tech-specific) -->
 | Attribute  | Type | Description  | Examples  | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) | Stability |
 |---|---|---|---|---|---|
-| [`db.name`](../attributes-registry/db.md) | string | The HBase namespace. [1] | `mynamespace` | `Conditionally Required` If applicable. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+| [`db.collection.name`](../attributes-registry/db.md) | string | The HBase table name. [1] | `mytable`; `ns:table` | `Conditionally Required` If applicable. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+| [`db.collection.namespace`](../attributes-registry/db.md) | string | The HBase namespace. [2] | `mynamespace` | `Conditionally Required` If applicable and is not default. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```
Comment (lmolkova) [inline docs/database/hbase.md:21]: Open question: hbase has default namespace, should we set to `default` or not set it?

File: docs/database/hbase.md [hunk 3]
```diff
@@ -17,9 +17,12 @@ described on this page.
...
 
-**[1]:** For HBase the `db.name` should be set to the HBase namespace.
+**[1]:** If table name includes the namespace, the `db.collection.name` SHOULD be set to the full table name.
+
+**[2]:** When performing table-related operations, the instrumentations SHOULD extract the namespace from the table name according to the [HBase table naming conventions](https://hbase.apache.org/book.html#namespace_creation).
```
Comment (lmolkova) [inline docs/database/hbase.md:25]: HBase tables can be named as `table` (default namespace) or `namespace:table`. 

Namespace is more than a naming convention e.g. it affects permissions or you can list tables within database.

However common operations (such as get/add rows) don't have namespace as a dedicated parameter, it's provided only as a part of a table name. See  https://hbase.apache.org/book.html#_using_rest_endpoints for the deatils..

Given it's easy to extract and there is no other way to obtain it, I added this note.

If we don't extract it, we won't have namespace as a separate attribute for common HBase operations. 

We'll also have to update how span name is constructed for HBase - it should include full table name and not the namespace.

File: docs/database/mongodb.md [hunk 4]
```diff
@@ -17,14 +17,17 @@ described on this page.
...
+| [`db.collection.name`](../attributes-registry/db.md) | string | The MongoDB collection being accessed within the database stated in `db.collection.namespace`. [1] | `public.users`; `customers` | `Required` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+| [`db.collection.namespace`](../attributes-registry/db.md) | string | The MongoDB database name. [2] | `customers`; `test.users` | `Conditionally Required` If applicable. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+| [`db.operation.name`](../attributes-registry/db.md) | string | The name of the command being executed. [3] | `findAndModify`; `getMore`; `update` | `Conditionally Required` [4] | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
 
 **[1]:** If the collection name is parsed from the query, it SHOULD match the value provided in the query and may be qualified with the schema and database name.
 
-**[2]:** See [MongoDB database commands](https://www.mongodb.com/docs/manual/reference/command/).
+**[2]:** <!-- TODO: overriding the base note, workaround for https://github.com/open-telemetry/build-tools/issues/299 -->
```
Comment (lmolkova) [inline docs/database/mongodb.md:26]: base note does not look great here. It's rendered as:

**[2]:**

File: docs/database/redis.md [hunk 5]
```diff
@@ -17,32 +17,32 @@ described on this page.
...
-| [`db.query.text`](../attributes-registry/db.md) | string | The full syntax of the Redis CLI command. [1] | `HMSET myhash field1 'Hello' field2 'World'` | `Recommended` [2] | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
-| [`network.peer.address`](../attributes-registry/network.md) | string | Peer address of the database node where the operation was performed. [3] | `10.1.2.80`; `/tmp/my.sock` | `Recommended` | ![Stable](https://img.shields.io/badge/-stable-lightgreen) |
+| [`db.collection.namespace`](../attributes-registry/db.md) | string | The index of the database being accessed as used in the [`SELECT` command](https://redis.io/commands/select). [1] | `0`; `1`; `15` | `Conditionally Required` If applicable. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```
Comment (lmolkova) [inline docs/database/redis.md:20]: Open question: java instrumentations use index if it's provided in the connection string. But it's dynamic and can be changed with `SELECT index` command. 
It's per-connection option and in theory can be cached and used until changed with `SELECT`

It's possible to get current index from redis via (`client info`)(https://redis.io/docs/latest/commands/client-info/https://redis.io/docs/latest/commands/client-info/) or `client list` commands, but it's a network call and can't be done on per-operation basis.

Given all of these, I think that some instrumentations may try to populate it, but then they have to support detecting and caching `SELECT index` commands.
If they don't it, then capturing index from the connection string seems unreliable.

**Comment by Oberon00** [comment]
Was there any discussion before this PR in some SIG meeting or similar? Is there any issue associated with this PR?

**Comment by lmolkova** [comment]
> Was there any discussion before this PR in some SIG meeting or similar? Is there any issue associated with this PR?

yes, this is the design approach discussed within DB semconv WG: https://docs.google.com/document/d/1zTi_Z4WBisytPnXTpwtnGOK-4Gv5oSTuMHvIQ8--5hg

It partially addresses #749, clarifies parts of #718, #714

File: docs/database/redis.md (hunk 5 above)
Comment (trask) [reply docs/database/redis.md:20]: > If they don't it, then capturing index from the connection string seems unreliable.

👍

if we need/want this in the future, it could be a new attribute e.g. `db.connection.default_namespace`

File: docs/database/hbase.md (hunk 2 above)
Comment (trask) [reply docs/database/hbase.md:21]: I'd suggest capturing it to avoid the question of whether missing value means default or not captured

**Comment by trask** [comment]
> if schema is provided as a part of table name (best practice), we have duplication

wdyt of just `SELECT {db.collection.name}` for the span name? I'm not sure we need `db.collection.namespace` in the span name, sort of like HTTP spans include the route but not the server name. you can always group span names further by `db.collection.namespace` if that's desired

**Comment by trask** [comment]
> where does the schema belong?

For SQL at least:

* `db.collection.name` will be whatever the user has in the query (may be partially or fully namespaced, or not namespaced at all)
* so the question is whether to capture schema in `db.collection.namespace` or not
  * it could be useful to capture in `db.collection.namespace`, since it gives you the "complete namespace" for the collection, which could be useful for grouping(?)
  * https://stackoverflow.com/questions/7942520/relationship-between-catalog-schema-user-and-database-instance refers to schema as a namespace, which makes me think that it may be more intuitive to include it in `db.collection.namespace`

File: docs/database/database-spans.md [hunk 6]
```diff
@@ -74,7 +74,7 @@ These attributes will usually be the same for all operations performed over the
...
 | [`db.collection.name`](../attributes-registry/db.md) | string | The name of a collection (table, container) within the database. [1] | `public.users`; `customers` | `Conditionally Required` [2] | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
-| [`db.name`](../attributes-registry/db.md) | string | This attribute is used to report the name of the database being accessed. For commands that switch the database, this should be set to the target database (even if the command fails). [3] | `customers`; `main` | `Conditionally Required` If applicable. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+| [`db.collection.namespace`](../attributes-registry/db.md) | string | The namespace containing database objects, fully qualified within the server address and port. [3] | `customers`; `test.users` | `Conditionally Required` If applicable. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```
Comment (gregkalapos) [inline docs/database/database-spans.md:77]: How should the implementation handle cases when a query touches multiple tables across multiple schemas?

MS SQL example with 2 schemas (`schema1` and `schema2`):

```
SELECT schema1.table1.C2
FROM schema1.table1
INNER JOIN schema2.table2 ON schema1.table1.C1 = schema2.table2.C1
```

1) To the "where should the schema be?" discussion: regardless of what we settle on, I think the spec should say something about queries with multiple schemas.
2) Not related to the PR, but shouldn't the `db.collection.name` say something about such cases as well? (specifically queries with multiple tables)

I think both multiple tables and multiple schemas can be handled the same (either choose 1 (e.g. from the SELECT clause). or comma separate, or something else, but do the same for `db.collection.namespace` and `db.collection.name`).

File: docs/attributes-registry/db.md [hunk 7]
```diff
@@ -21,16 +19,17 @@
...
+| `db.collection.namespace` | string | The namespace containing database objects, fully qualified within the server address and port. [2] | `customers`; `test.users` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
 | `db.instance.id` | string | An identifier (address, unique name, or any other identifier) of the database instance that is executing queries or mutations on the current connection. This is useful in cases where the database is running in a clustered environment and the instrumentation is able to record the node executing the query. The client may obtain this value in databases like MySQL using queries like `select @@hostname`. | `mysql-e26b99z.example.com` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
-| `db.name` | string | This attribute is used to report the name of the database being accessed. For commands that switch the database, this should be set to the target database (even if the command fails). [2] | `customers`; `main` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```
Comment (jcocchi) [inline docs/attributes-registry/db.md:25]: Is there a way we can add `db.collection.namespace` without getting rid of `db.name` entirely? I recognize there are gaps with `db.name` for some systems, but this is an intuitive name that covers the simple use case well.

Comment (trask) [reply docs/attributes-registry/db.md:25]: sure, I think it's worth reconsidering the name `db.collection.namespace` given we are no longer thinking to parse the namespace from `db.collection.name`

in Java, for example, `db.collection.namespace` will probably be captured via some combination of

  * [java.sql.Connection#getCatalog()](https://docs.oracle.com/javase/8/docs/api/java/sql/Connection.html#getCatalog%E2%80%93)
  * [java.sql.DatabaseMetaData.getUserName()](https://docs.oracle.com/javase/8/docs/api/java/sql/DatabaseMetaData.html#getUserName--)
    * in some databases (e.g. PostgreSQL and Oracle), users are owners of database objects and so they are part of an object's namespace
  * [DatabaseMetaData.getURL()](https://docs.oracle.com/javase/8/docs/api/java/sql/DatabaseMetaData.html#getURL--)
    * this could be used to parse out some parts of the "namespace", e.g. the Oracle Instance Name

and these aren't really scoped to the collection in `db.collection.name`, so maybe we shouldn't scope namespace to `db.collection.*`

File: docs/database/database-spans.md (hunk 6 above)
Comment (trask) [reply docs/database/database-spans.md:77]: > Not related to the PR, but shouldn't the `db.collection.name` say something about such cases as well? (specifically queries with multiple tables)

I believe it does, check out the footnote for `db.collection.name` in section https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/database-spans.md#common-attributes

> [2]: If readily available. Otherwise, if the instrumentation library parses `db.query.text` to capture `db.collection.name`, then it SHOULD be the first collection name found in the query.

did we maybe miss this somewhere else?

File: docs/attributes-registry/db.md (hunk 7 above)
Comment (trask) [reply docs/attributes-registry/db.md:25]: @jcocchi do I understand correctly, for CosmosDB there are 3 "layers"
* Account - this would be captured in `server.address`?
* Database - this would be captured in the "namespace" attribute we're discussing?
* Collection - this would be captured in `db.collection.name`?

Comment (trask) [reply docs/attributes-registry/db.md:25]: > * [java.sql.Connection#getCatalog()](https://docs.oracle.com/javase/8/docs/api/java/sql/Connection.html#getCatalog%E2%80%93)
> * [java.sql.DatabaseMetaData.getUserName()](https://docs.oracle.com/javase/8/docs/api/java/sql/DatabaseMetaData.html#getUserName--)

btw, I don't know
 if these make network calls, and so whether they would be practical to use from instrumentation

Comment (lmolkova) [reply docs/attributes-registry/db.md:25]: The catalog seems to be static and local:
- postgres [1](https://github.com/pgjdbc/pgjdbc/blob/d5cfb04d1413685ce4d4a9a2f641d88b514bcdaf/pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java#L1115) and [2](https://github.com/pgjdbc/pgjdbc/blob/d5cfb04d1413685ce4d4a9a2f641d88b514bcdaf/pgjdbc/src/main/java/org/postgresql/core/QueryExecutorBase.java#L141))
- [SQL Server](https://github.com/microsoft/mssql-jdbc/blob/193edbdaf085786fbe47227e80b88d942e0547ad/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java#L4900)
- [MySQL](https://github.com/mysql/mysql-connector-j/blob/805f872a57875f311cb82487efcfb070411a3fa0/src/main/user-impl/java/com/mysql/cj/jdbc/ConnectionImpl.java#L1088)
- Oracle: gave up reading decompiled jar

User - mostly dynamic and requires network call:
- [SQL Server](https://github.com/microsoft/mssql-jdbc/blob/main/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDatabaseMetaData.java#L1784) - dynamic
- [Postgres](https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java#L1671) - dynamic
- [MySQL](https://github.com/mysql/mysql-connector-j/blob/805f872a57875f311cb82487efcfb070411a3fa0/src/main/user-impl/java/com/mysql/cj/jdbc/DatabaseMetaData.java#L4175) - sometimes static sometimes dynamic depending on the config. Seems to be dynamic by default

Schema - mostly dynamic and requires network call:
- [Postgres](https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java#L1672) - dynamic
- [SQL Server](https://github.com/microsoft/mssql-jdbc/blob/193edbdaf085786fbe47227e80b88d942e0547ad/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java#L7271) - dynamic
- [MySQL](https://github.com/mysql/mysql-connector-j/blob/805f872a57875f311cb82487efcfb070411a3fa0/src/main/user-impl/java/com/mysql/cj/jdbc/ConnectionImpl.java#L2393) static/local and the same as database name (depends on mode)

Comment (trask) [reply docs/attributes-registry/db.md:25]: For Oracle JDBC driver:

Catalog - [always null](https://github.com/caot/ojdbc6-11.2.0.2.0.src/blob/41eef5e34775532c7aec2f03d41d360152676e00/src/oracle/jdbc/driver/PhysicalConnection.java#L3230-L3234)

User - [cached after the first call](https://github.com/caot/ojdbc6-11.2.0.2.0.src/blob/41eef5e34775532c7aec2f03d41d360152676e00/src/oracle/jdbc/driver/PhysicalConnection.java#L4042-L4049)

Comment (trask) [reply docs/attributes-registry/db.md:25]: I think this part of https://stackoverflow.com/questions/7942520/relationship-between-catalog-schema-user-and-database-instance is interesting:

to identify any object in running server, you need

* Oracle - schema name + object name
* PostgreSQL - database name + schema name + object name
* MySQL - database name + object name
* SQLServer - database name + schema name + object name

to identify the running server itself

* Oracle - [instance name](https://docs.oracle.com/en/database/oracle/oracle-database/21/jajdb/)
* PostgreSQL - n/a
* MySQL - n/a
* SQLServer - [instance name](https://learn.microsoft.com/en-us/sql/connect/jdbc/building-the-connection-url?view=sql-server-ver16)

so we have these 4 things:

* instance name - doesn't change after connection, can be captured from connection string (probably maps to `db.instance.id`, maybe could consider renaming to `db.instance.name`)
* database name - can change after connection, JDBC drivers seem to track these changes and expose via getCatalog, but could be harder to capture in some languages
* schema name - can change after connection, JDBC drivers don't appear to track these changes, so probably cannot capture in the generic case
* object name - `db.collection.name`

Comment (trask) [reply docs/attributes-registry/db.md:25]: it doesn't seem that the above JDBC drivers track "use <database>" calls, and so getCatalog() could be incorrect, though I guess it could be said that best practice is to call `setCatalog(<database>)`, which the drivers do track

File: docs/database/database-spans.md (hunk 6 above)
Comment (gregkalapos) [reply docs/database/database-spans.md:79]: Oh, indeed, sorry (I'd have expected that part under footnote [1] - but maybe it's just me, doesn't matter). 

So cases with multiple tables seem to be handled fine. So if we decide to capture schema, I suggest we have something similar for that as well.

File: docs/attributes-registry/db.md (hunk 7 above)
Comment (lmolkova) [reply docs/attributes-registry/db.md:25]: I updated the note on the SQL namespace to be

```yaml
- ref: db.collection.namespace
  brief: The namespace containing database objects, fully qualified within the server address and port.
  note: >
    Namespace contains different identifiers depending on the database system:

    * MySQL - database or schema name. For example, `products.dbo`
    * PostgreSQL - database and schema name. For example, `products.public`
    * MS SQL Server - instance name, database and schema name. For example, `test.products.public`
    * Oracle - instance name and schema name. For example, `test.products`

    Current database name can usually be obtained using database driver API such as
    [JDBC `Connection.getCatalog()`](https://docs.oracle.com/javase/8/docs/api/java/sql/Connection.html#getCatalog--)
    or [.NET `SqlConnection.Database`](https://learn.microsoft.com/dotnet/api/system.data.sqlclient.sqlconnection.database).

    If instrumentation cannot reliably determine the database name (for example, if database can be changed
    in runtime without instrumentation being aware of it), it SHOULD NOT set `db.collection.namespace`.

    Instrumentations that parse SQL statements MAY use database name provided in the connection string
    and track the currently selected database name as long as current database name is associated with
    current connection and can't be changed without instrumentation being aware of it.
 
    For commands that switch the database, this should be set to the target database (even if the command fails).

    TODO schema.
```

From here, I see the following options:
1. allow to skip schema if it's not known.
    - i.e. MS SQL Server would have `db.collection.namespace = instance.db` ~~most~~all of the time. 
    - Oracle might need the whole attribute to be opt-in to make a network call to get the current schema and then monitor `ALTER SESSION SET CURRENT_SCHEMA` calls
2. don't include schema into the namespace at all. 
    - it could be bad for Oracle, but all options are bad for it.
3. Gave up on the idea to have namespace attribute and just do cosmetic improvements to the DB name. I still see the value in the namespace hirarchy though:
    - place for the instance-within-host identifier
    - possible place for [cluster](https://github.com/open-telemetry/semantic-conventions/issues/858)/shard/partition-id 

Let's talk more about it on the WG call.

**Comment by gregkalapos** [comment]
As discussed in the last WG call, looked into what we can collect in .NET.

The `DbConnection` ADO interface has following relevant methods:
- `DbConnection.DataSource`: name of the database server
- `DbConnection.Database`: name of the database
- `DbConnection.GetSchema(string)`: returns a `DataTable` which contains metadata info about the objects in the db including schema and collection. [Here](https://learn.microsoft.com/en-us/dotnet/framework/data/adonet/getschema-and-schema-collections) is a super useful article on this.

I ran following code snippet with different DBs and libraries:

```
static void DisplayConnectionData(DbConnection connection)
{
    Console.WriteLine($"DataSource: {connection.DataSource}, Database: {connection.Database}");
    Console.WriteLine("============================");

    DataTable table = connection.GetSchema("Tables");
    foreach (System.Data.DataRow row in table.Rows)
    {
        foreach (System.Data.DataColumn col in table.Columns)
        {
            Console.WriteLine("{0} = {1}", col.ColumnName, row[col]);
        }
        Console.WriteLine("============================");
    }
}
```

Above code queries the `Tables` schema collection (so it only returns infos about tables). [Here](https://mysqlconnector.net/overview/schema-collections/) is a list about other collections (e.g. `Procedures` could be used for stored procedures).

## MS SQL Server with [System.Data.SqlClient](https://www.nuget.org/packages/System.Data.SqlClient)

Output:

```
DataSource: 127.0.0.1,50312, Database: master
============================
TABLE_CATALOG = master
TABLE_SCHEMA = dbo
TABLE_NAME = spt_fallback_db
TABLE_TYPE = BASE TABLE
============================
....similar info for all tables...
```

## MySQL with [MySqlConnector](https://www.nuget.org/packages/MySqlConnector)

Output:

```
DataSource: 127.0.0.1, Database: test
============================
TABLE_CATALOG = def
TABLE_SCHEMA = information_schema
TABLE_NAME = CHARACTER_SETS
TABLE_TYPE = SYSTEM VIEW
ENGINE =
VERSION = 10
ROW_FORMAT =
TABLE_ROWS = 0
AVG_ROW_LENGTH = 0
DATA_LENGTH = 0
MAX_DATA_LENGTH = 0
INDEX_LENGTH = 0
DATA_FREE = 0
AUTO_INCREMENT =
CREATE_TIME = 12.04.2024 14:18:11
UPDATE_TIME =
CHECK_TIME =
TABLE_COLLATION =
CHECKSUM =
CREATE_OPTIONS =
TABLE_COMMENT =
============================
....similar info for all tables...
```

## PostgreSql with [Npgsql](https://www.nuget.org/packages/Npgsql)

Output:

```
DataSource: tcp://127.0.0.1:54251, Database: postgres
============================
table_catalog = postgres
table_schema = public
table_name = foo
table_type = BASE TABLE
============================
```

So, overall it seems, at least in ADO.NET we can get the schema. Issue I see with this is that this seems to query the database - not only that, but ` connection.GetSchema("Tables")` returns the schema info for all tables, which can be very high overhead.

File: docs/attributes-registry/db.md [hunk 8]
```diff
@@ -21,16 +19,27 @@
...
 **[1]:** If the collection name is parsed from the query, it SHOULD match the value provided in the query and may be qualified with the schema and database name.
 
-**[2]:** In some SQL databases, the database name to be used is called "schema name". In case there are multiple layers that could be considered for database name (e.g. Oracle instance name and schema name), the database name to be used is the more specific layer (e.g. Oracle schema name).
+**[2]:** For many database systems, the namespace matches the database (catalog, keyspace, schema) name.
+In general case, however, namespace consist of several qualifiers:
+
+* Instance name - MS SQL Server or Oracle, allow to host multiple database engines on the same host.
+* Database name - The name of the database, keyspace, namespace, etc.
+* Schema name - such as PostgreSQL or MS SQL Server schemas.
+* Other sub-namespaces which may include partition or shard identifiers.
```
Comment (jcocchi) [inline docs/attributes-registry/db.md:37]: My understanding is that `db.collection.namespace` captures the hierarchy "above" the collection to fully qualify it. The partition/ shard would be "below" the collection name in hierarchy, so I don't think it makes sense to include those identifiers here.

File: model/trace/database.yaml [hunk 9]
```diff
@@ -232,8 +269,33 @@ groups:
...
         tag: tech-specific
+      - ref: db.collection.namespace
+        brief: The namespace containing database objects, fully qualified within the server address and port.
+        note: >
+          Namespace contains different identifiers depending on the database system:
+
+          * MySQL - database or schema name. For example, `products.dbo`
+          * PostgreSQL - database and schema name. For example, `products.public`
+          * MS SQL Server - instance name, database and schema name. For example, `test.products.public`
+          * Oracle - instance name and schema name. For example, `test.products`
```
Comment (jcocchi) [inline model/trace/database.yaml:282]: Adding Cosmos DB and MongoDB. One note for MongoDB- "namespace" refers to the fully qualified collection name *including* the collection name itself, but given the definition of this attribute collection name wouldn't be included.

```suggestion
          * Oracle - instance name and schema name. For example, `test.products`
          * Cosmos DB - database name. For example, `products`
          * MongoDB - database name. For example, `products`
```

File: docs/attributes-registry/db.md (hunk 7 above)
Comment (jcocchi) [reply docs/attributes-registry/db.md:25]: Added examples for Cosmos and Mongo inline

          * Cosmos DB - database name. For example, `products`
          * MongoDB - database name. For example, `products`

File: model/registry/db.yaml [hunk 10]
```diff
@@ -203,30 +203,25 @@ groups:
...
+
+          * Instance name - MS SQL Server or Oracle, allow to host multiple database engines on the same host.
+          * Database name - The name of the database, keyspace, namespace, etc.
+          * Schema name - such as PostgreSQL or MS SQL Server schemas.
+          * Other sub-namespaces which may include partition or shard identifiers.
+
+          Namespace SHOULD include all such identifiers applicable to the database system and available to the instrumentation.
+
+          Semantic conventions for individual database systems SHOULD document what `db.collection.namespace`
+          means in the context of that system.
```
Comment (trask) [inline model/registry/db.yaml:223]: my thought is to try to stick to general statements here and then lean on

> Semantic conventions for individual database systems SHOULD document what `db.collection.namespace` means in the context of that system.

```suggestion
          If a database system has multiple namespace components, they should be concatenated (potentially using database system specific conventions) from most general to most specific namespace component, and more specific namespaces should not be captured without the more general namespaces, to ensure that "startswith" queries for the more general namespaces will be valid.

          Semantic conventions for individual database systems SHOULD document what `db.collection.namespace`
          means in the context of that system.
```

File: docs/attributes-registry/db.md (hunk 7 above)
Comment (trask) [reply docs/attributes-registry/db.md:25]: any objection to marking this discussion as resolved?

File: model/trace/database.yaml [hunk 11]
```diff
@@ -33,6 +30,12 @@ groups:
           conditionally_required: >
             If readily available. Otherwise, if the instrumentation library parses `db.query.text` to capture
             `db.collection.name`, then it SHOULD be the first collection name found in the query.
+        note: >
+          If the collection name is parsed from the query, it SHOULD match the value provided
+          in the query and may be qualified with the schema and database name.
```
Comment (jack-berg) [inline model/trace/database.yaml:35]: #nit: Is this note related to this PR?

**Review by jack-berg — APPROVED** [review]

**Review by gregkalapos — APPROVED** [review]

**Review by alanwest — APPROVED** [review]

File: docs/attributes-registry/db.md (hunk 7 above)
Comment (lmolkova) [reply docs/attributes-registry/db.md:28]: Based on the previous discussion (Wed 4/17 DB WG meeting) with @jcocchi I believe `db.namespace` was a good solution and we can resolve this discussion.

File: model/trace/database.yaml (hunk 11 above)
Comment (lmolkova) [reply model/trace/database.yaml]: Great catch, it's redundant since it matches the note in the registry

File: docs/attributes-registry/db.md (hunk 8 above)
Comment (lmolkova) [reply docs/attributes-registry/db.md]: removed

File: model/trace/database.yaml (hunk 9 above)
Comment (lmolkova) [reply model/trace/database.yaml]: thank you! Reflected in the `db.namespace` brief overrides on Cosmos and Mongo

**Review by jsuereth — APPROVED** [review]

File: model/registry/db.yaml [hunk 12]
```diff
@@ -13,19 +13,20 @@ groups:
...
-          the database name to be used is the more specific layer (e.g. Oracle schema name).
-        examples: [ 'customers', 'main' ]
+          If a database system has multiple namespace components, they SHOULD be concatenated
+          (potentially using database system specific conventions) from most general to most
+          specific namespace component, and more specific namespaces SHOULD NOT be captured without
+          the more general namespaces, to ensure that "startswith" queries for the more general namespaces will be valid.
+
+          Semantic conventions for individual database systems SHOULD document what `db.namespace`
+          means in the context of that system.
+        examples: [ 'customers', 'test.users' ]
```
Comment (Oberon00) [inline model/registry/db.yaml:29]: These sound like table (collection) names rather than namespace (database) names
